### ISSUE TITLE
Regression in Socket#close - EBADF after unsuccessful connection

### ISSUE THREAD

**Issue by p0deje** [jruby/jruby#7483]
The issue was originally reported in #5709 and successfully fixed in JRuby 9.2.8. However, the regression was introduced in JRuby 9.3 and is still present in JRuby 9.4.

I cannot reopen original bug - hence new issue.

**Environment Information**
  
Provide at least:

* JRuby 9.4.0.0
* macOS 13 arm64

Other relevant info you may wish to add:
* Installed or activated gems
* Application/framework version (e.g. Rails, Sinatra)
* Environment variables

**Expected Behavior**

Run the following script:

```ruby
require 'socket'

sock = Socket.new(Socket::AF_INET, Socket::SOCK_STREAM)
addr = Socket.pack_sockaddr_in(9250, '127.0.0.1')

begin
  sock.connect_nonblock(addr)
rescue Errno::EINPROGRESS
  IO.select(nil, [sock], nil, 1) # select socket for writing
  sock.getsockopt(Socket::SOL_SOCKET, Socket::SO_ERROR) # read any errors from last connection
  retry
rescue Errno::ECONNREFUSED, Errno::EINVAL
  puts 'Open!'
end

sock.close
```

This code works fine on MRI 3.0:

```
$ ruby socket.rb
Open!
```

**Actual Behavior**

The same code fails on JRuby 9.4:

```
$ ruby socket.rb
Open!
Errno::EBADF: Bad file descriptor - No message available
   Unable to find org/jruby/ext/socket/RubySocket.java to read failed line
  <main> at socket.rb:16
```

**Comment by headius** [comment]
This is a little bit confusing. The socket is not open, so attempting to close it should (in my mind) produce an error. But on MRI, it does not, so it seems like they must be silently ignoring the error.

**Comment by mohamedhafez** [comment]
The difference between MRI and JRuby does cause some minor issues, for example I've been running this patch in production, although I can't seem to get a response when trying to get it into the official net/http gem: https://github.com/ruby/net-http/pull/45

**Comment by headius** [comment]
@p0deje @mohamedhafez I think I figured out the magic juice: https://github.com/jruby/jruby/pull/7635

When a `connect_nonblock` truly fails (e.g. connection refused, no route to host, other fatal errors) we need to clean up the IO state so it does not look open. Calling this cleanup method nulls out the file descriptor, so that calls to `close` will see it is already closed and do nothing.

If you folks can test that PR branch and make sure it works for you, I'd appreciate it!

**Comment by mohamedhafez** [comment]
@headius awesome! Using the debugger and breakpointing midway through the process of creating the connection to a local server, then killing the local server, I was able to recreate the `Errno::EBADF` error in 9.4.1.0, but then got the correct behavior once I applied your patch, and got the original `Errno::ECONNREFUSED` exception.

I'm also running the patch in production, and will report back if there are any issues, but I'm guessing we are good!
